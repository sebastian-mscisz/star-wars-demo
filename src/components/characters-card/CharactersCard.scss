@import 'src/variables';

.sw-characters-card {
    width: calc((100% / var(--columns)) - var(--gap) + (var(--gap) / var(--columns)));

    &__tile {
        background-color: $base-background-color-light;
        border-radius: 10%;
        border: 2px solid $base-color;

        &__link {
            display: flex;
            flex-wrap: wrap;
            text-decoration: none;

            &__image-wrapper {
                width: 100%;
                border-radius: 10% 10% 0 0;
                overflow: hidden;
                border-bottom: 3px solid $base-color;
                transition: all 0.3s linear;
                aspect-ratio: 8/11;
                background-color: $base-background-color;

                &__display-spinner {
                    height: 100%;
                    display: flex;
                    justify-content: center;
                }

                &__hide-spinner {
                    display: none;
                }

                &__hide-img {
                    visibility: hidden;
                }

                & img {
                    width: 100%;
                    height: 100%;
                    transition: all 0.3s linear;
                }
            }

            &__text {
                width: 100%;
                text-align: center;
                color: $base-color;
                transition: all 0.3s linear;

                &:hover {
                    color: $base-color-light;
                }
            }
        }
    }

    &:hover &__tile__link__image-wrapper {
        border-bottom: 3px solid $base-color-light;

        & img {
            transform: scale(1.2);
        }
    }

    &:hover &__tile__link__text {
        color: $base-color-light;
    }
}