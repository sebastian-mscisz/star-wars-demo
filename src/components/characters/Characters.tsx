import { withHandlers } from 'src/hoc';
import { CharactersUi } from './ui/CharactersUi';
import { useCharacters } from 'src/hooks';

const CharactersWithHandlers = withHandlers(CharactersUi);

export const Characters: React.FC = () => {
    const { data: characters, isLoading, loadingError } = useCharacters();

    return <CharactersWithHandlers isLoading={isLoading} characters={characters} hasError={!!loadingError} />;
};
