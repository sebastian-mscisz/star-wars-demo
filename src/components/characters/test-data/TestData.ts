import { Characters, Character, Species } from 'src/common';

export const TestCharacter: Character = {
    name: 'Character1',
    url: 'https://swapi.info/api/people/1',
    homeworld: 'https://swapi.info/api/planets/1',
    species: ['species/1'],
    vehicles: ['https://swapi.info/api/vehicles/1'],
};

export const TestCharacters: Characters = [
    TestCharacter,
    {
        name: 'Character2',
        url: 'https://swapi.info/api/people/2',
        vehicles: ['https://swapi.info/api/vehicles/1', 'https://swapi.info/api/vehicles/2'],
        homeworld: 'https://swapi.info/api/planets/2',
        species: ['species/2'],
    },
    {
        name: 'Character3',
        url: 'https://swapi.info/api/people/3',
        vehicles: [],
        homeworld: 'https://swapi.info/api/planets/3',
        species: ['species/3'],
    },
    {
        name: 'Character4',
        url: 'https://swapi.info/api/people/4',
        vehicles: ['https://swapi.info/api/vehicles/3'],
        homeworld: 'https://swapi.info/api/planets/4',
        species: ['species/4'],
    },
    {
        name: 'Character5',
        url: 'https://swapi.info/api/people/5',
        vehicles: ['https://swapi.info/api/vehicles/4', 'https://swapi.info/api/vehicles/5'],
        homeworld: 'https://swapi.info/api/planets/4',
        species: ['species/5'],
    },
];

export const TestSpecies: Species = [
    {
        name: 'Species2',
        url: 'species/2',
    },
    {
        name: 'Species2',
        url: 'species/2',
    },
    {
        name: 'Species3',
        url: 'species/3',
    },
    {
        name: 'Species4',
        url: 'species/4',
    },
    {
        name: 'Species5',
        url: 'species/5',
    },
];
