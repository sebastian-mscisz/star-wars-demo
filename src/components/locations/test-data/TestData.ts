import { Locations, Location } from 'src/common';

export const TestLocation: Location = {
    name: 'Location1',
    url: 'https://swapi.info/api/planets/1',
    residents: ['https://swapi.info/api/people/1'],
    population: '1000',
};

export const TestLocations: Locations = [
    TestLocation,
    {
        name: 'Location2',
        url: 'https://swapi.info/api/planets/2',
        residents: ['https://swapi.info/api/people/2'],
        population: '5343',
    },
    {
        name: 'Location3',
        url: 'https://swapi.info/api/planets/3',
        residents: ['https://swapi.info/api/people/3'],
        population: '13333',
    },
    {
        name: 'Location4',
        url: 'https://swapi.info/api/planets/4',
        residents: ['https://swapi.info/api/people/4', 'https://swapi.info/api/people/5'],
        population: '9999',
    },
    {
        name: 'Location5',
        url: 'https://swapi.info/api/planets/5',
        residents: [],
        population: '999',
    },
];
