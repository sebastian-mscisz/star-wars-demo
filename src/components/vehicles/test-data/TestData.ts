import { Vehicles, Vehicle } from 'src/common';

export const TestVehicle: Vehicle = {
    name: 'Vehicle1',
    url: 'https://swapi.info/api/vehicles/1',
    vehicle_class: 'Air speeder',
    pilots: ['character/1', 'character/2'],
};

export const TestVehicles: Vehicles = [
    TestVehicle,
    {
        name: 'Vehicle2',
        url: 'https://swapi.info/api/vehicles/2',
        pilots: ['character/2'],
        vehicle_class: 'Land speeder',
    },
    {
        name: 'Vehicle3',
        url: 'https://swapi.info/api/vehicles/3',
        pilots: ['character/4'],
        vehicle_class: 'Water speeder',
    },
    {
        name: 'Vehicle4',
        url: 'https://swapi.info/api/vehicles/4',
        pilots: ['character/5'],
        vehicle_class: 'Fire speeder',
    },
    {
        name: 'Vehicle5',
        url: 'https://swapi.info/api/vehicles/5',
        pilots: ['character/5'],
        vehicle_class: 'Air speeder',
    },
];
