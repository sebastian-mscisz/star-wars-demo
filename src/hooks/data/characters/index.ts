export * from './useCharacters';
export * from './useCharacterDetails';
