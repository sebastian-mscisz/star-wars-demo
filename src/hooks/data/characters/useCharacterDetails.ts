import { useMemo } from 'react';
import { CharacterWithNeighbours } from 'src/common';
import { TDataState, CharactersDataSelector, fetchCharacters, useGetApiSingleData } from 'src/store';
import { findItemByIdWithNeighbors } from '../../utils';

export const useCharacterDetails = (characterId: string): TDataState<CharacterWithNeighbours | undefined> => {
    const characterFilter = useMemo(
        () => findItemByIdWithNeighbors('https://swapi.info/api/people/' + characterId),
        [characterId],
    );

    return useGetApiSingleData(fetchCharacters, CharactersDataSelector, characterFilter);
};
