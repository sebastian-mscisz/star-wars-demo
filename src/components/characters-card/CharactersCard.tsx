import React from 'react';
import { Character, characterDetailsPagePath, getIdFromUrl } from 'src/common';
import { NavLink } from 'react-router-dom';
import './CharactersCard.scss';

type CharactersCardProps = {
    item?: Character;
};

export const CharactersCard: React.FC<CharactersCardProps> = ({ item }) => {
    if (!item) return null;

    const { name, url } = item;

    return (
        <div className="sw-characters-card">
            <div className="sw-characters-card__tile">
                <NavLink
                    className={'sw-characters-card__tile__link'}
                    key={name}
                    to={characterDetailsPagePath(getIdFromUrl(url))}
                >
                    <div className="sw-characters-card__tile__link__image-wrapper">
                        <img loading="lazy" src={`/characters/${getIdFromUrl(url)}.jpg`} alt="Star wars character" />
                    </div>
                    <h5 className="sw-characters-card__tile__link__text">{name}</h5>
                </NavLink>
            </div>
        </div>
    );
};
