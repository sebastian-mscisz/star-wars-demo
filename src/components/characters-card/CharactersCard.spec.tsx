import { BrowserRouter } from 'react-router-dom';
import { CharactersCard } from './CharactersCard';
import { render } from '@testing-library/react';
import { TestCharacter } from '../characters/test-data';

describe('Characters card', () => {
    it('should render', () => {
        const { asFragment } = render(<CharactersCard item={TestCharacter} />, { wrapper: BrowserRouter });
        expect(asFragment()).toMatchSnapshot();
    });
});
