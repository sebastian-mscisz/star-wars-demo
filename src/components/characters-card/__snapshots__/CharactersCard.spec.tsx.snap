// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Characters card should render 1`] = `
<DocumentFragment>
  <div
    class="sw-characters-card"
  >
    <div
      class="sw-characters-card__tile"
    >
      <a
        class="sw-characters-card__tile__link"
        href="/character/1"
      >
        <div
          class="sw-characters-card__tile__link__image-wrapper"
        >
          <img
            alt="Star wars character"
            loading="lazy"
            src="./characters/1.jpg"
          />
        </div>
        <h5
          class="sw-characters-card__tile__link__text"
        >
          Character1
        </h5>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
